### ISSUE TITLE
Give more flexibility with AND and OR logic in tag or list selection and inversion.

### ISSUE THREAD

**Issue by Alphalpha007** [mpcjanssen/simpletask-android#498]

**Comment by mpcjanssen** [comment]
Inversion can be done by clicking on the tag or list blue bar header in the left drawer. Anything more exotic than this can be done in Lua. I don't want to complicate the UI even more.

I am open for brilliant UI ideas though.

**Comment by Alphalpha007** [comment]
I'm glad it's in AND mode now.

**Comment by Alphalpha007** [comment]
Oh it's not. It's in OR mode.

**Comment by smichel17** [comment]
I think the biggest improvement will be if we separate saved filters from quick filtering via the left drawer.

I don't think we should change anything else about filtering until we do that. Then we can re-evaluate filter changes afterwards.

I suspect that it will make sense to make saved filters slightly more configurable (eg, by adding AND), and for quick filtering, to use OR for tags, OR if there is only one list selected, and ONLY if multiple lists are selected.

**Comment by Alphalpha007** [comment]
Looks like this will just make things complicated. Lua is the way to go. Closing issue.

**Comment by smichel17** [comment]
I'll note that I don't think we should add any logic to saved filters. It should be along the lines of how it is now: you either pick OR mode or INVERT mode and then select lists/tags. I'm suggesting that it *may* make sense to add an AND mode there.

....but again, that we shouldn't worry about that before separating filtering into two stages.

**Comment by doranism** [comment]
> Inversion can be done by clicking on the tag or list blue bar header

@mpcjanssen I didn't know that, although I am using Simpletask for two years now :-) It would help, if the function would be more vivible in some way.

**Comment by mpcjanssen** [comment]
@doranism it's more discoverable in the filter activity (the Invert filter checkbox). The click on title is just a cute shortcut.

**Comment by t214c** [comment]
Hi,

I really need this feature (AND).

Meanwhile, if  this can be done in Lua, could you provide a script for enabling this on any tags/lists combination? i.e. if multiple tags/lists are selected, show tasks with *both* not either.

Thanks :)

**Comment by mpcjanssen** [comment]
@t214c lua doesn't enable this, you should write a custom Lua filter. If you need help please log a separate issue asking for help with details.

**Comment by t214c** [comment]
You mean custom for each possible configuration?

**Comment by smichel17** [comment]
> I am open for brilliant UI ideas though.

Currently you can tap the List/Tag header to invert it. What it tapping it a 3rd time swapped it to ALL? So the 3 options would be: ANY / ALL / NONE. That should be relatively easy to implement -- very little new UI.

**Comment by mpcjanssen** [comment]
I like that, would be even better with some sort of indication of which mode is active.

**Comment by mpcjanssen** [comment]
Maybe text indication is enough.

**Comment by t214c** [comment]
> 
> 
> I like that, would be even better with some sort of indication of which mode is active.

Seconded. I suggest:
`Tasks with any of these tags:`
`Tasks with all of these tags:`
`Tasks without these tags:`

or something similar.

Can a default behavior be set in the settings, please? :)

And since we are on the topic, how about Showing how many tasks under each tag/list, [Like this](https://f-droid.org/repo/me.cgarcia.yarr/en-US/phoneScreenshots/screenshot_phone_2.png).

Thanks!

**Comment by mpcjanssen** [comment]
@t214c please don't add additional requests or topics to an issue. It makes it difficult to track progress. See #995

**Comment by t214c** [comment]
OK. Sorry.

**Comment by mpcjanssen** [comment]
@t214c no problem, keep the requests coming. It's much appreciated.

**Comment by smichel17** [comment]
> Maybe text indication is enough.

I think we should start with text indication. We'd also need to change the inversion checkbox in the filter screen to a drop-down.

Making the functionality more discoverable is a separate problem -- present in the current UI whether or not we add "AND" sorting.

**Comment by mpcjanssen** [comment]
One challenge left is how to implement this without breaking existing filters. Most natural would be a new field in the filter json which gets filled for existing filters based on the invert value.

**Comment by smichel17** [comment]
> Most natural would be a new field in the filter json which gets filled for existing filters based on the invert value.

Agreed. We should probably also continue including the old key (when one of the existing options is in use), so that filters exported from a new version can still be imported into an old version. Then in a couple versions when most people have upgraded we can drop it. Doesn't affect parsing logic (try new key, fall back to old key), so it's not that much additional work.

**Comment by mpcjanssen** [comment]
Old to new is not a requirement IMO.

**Comment by smichel17** [comment]
Not a requirement but it's also only a couple lines of code.

```
if type != ALL {
    invert = type != any
    old way to store(invert)
}
```

**Comment by t214c** [comment]
> 
> 
> @t214c no problem, keep the requests coming. It's much appreciated.

Well, I commented at #41. Not sure if you saw it.

I have a few more but didn't want to add too many at a time, but since you asked :P

**Comment by smichel17** [comment]
> partial filter

No, but a while back we discussed separating filtering into two steps: the "saved filter" (right drawer, changeable by editing it), which we decided to call a *query*, and then the "quick filter" (left drawer, changeable via the gui). The left drawer would only show lists and tags of tasks that were selected by the query. Except for that last part, you can achieve a similar effect today by doing filtering in saved filters exclusively through lua. I think it's a big improvement if you have a big todo.txt file (or especially, if it has multiple types of things with different sorts of tags, like a mix of tasks and shopping/reading lists).

> not *"learn-a-new-language"* advanced

I will mention, lua is one of the easiest languages to learn, especially if you're not doing anything fancy. It has a single data structure (dictionaries). If you've done any programming at all, you should be able to learn what you need within an hour or two, especially given examples (which ST unfortunately does not provide; I've shared my configuration before). There's a free online book, "Programming in Lua". Read it from the start until you have enough (just the first few seconds, unless you want to do something really complicated).

**Comment by t214c** [comment]
@smichel17 

Thanks for the info. Lua is now fairly high on my things to learn. Still, no matter how easy, it can't compare with 1-2-3 taps!

Borrowing from SQL, saved filters might be called "views".

Are you saying that I can apply quick partial filtering using Lua scripts. For example, I apply filter A (which changes sorting to state X), then apply a Lua filter B (which doesn't affect sorting at all)?

**Comment by t214c** [comment]
@smichel17 

Thanks for the info. Lua is now fairly high on my things to learn. Still, no matter how easy, it can't compare with 1-2-3 taps!

Borrowing from SQL, saved filters might be called "views".

Are you saying that I can apply quick partial filtering using Lua scripts. For example, I apply filter A (which changes sorting to state X), then apply a Lua filter B (which doesn't affect sorting at all)?


@mpcjanssen Any possibility that you give the suggestion [here ](#issuecomment-953454523)and [here](707#issuecomment-954339146) a try?

**Comment by smichel17** [comment]
IIRC _views_ was one of the ideas for what to call them originally (and was where the inspiration for "queries" came from). I don't remember why we decided not to go with it.

I'm saying you can use lua to filter to only tasks with `+a`, then use the left drawer to select `+b` and `+c`, in order to get tasks with `+a AND (+b OR +c)`.

**Comment by mpcjanssen** [comment]
@t214c I don't really have the time to do any serious development right now. Especially UI changes because they somehow always end up as a complicated mess on Android. However PRs are always welcome.

**Comment by mpcjanssen** [comment]
BTW: I think the multistate boxes for filtering is too complex.

**Comment by t214c** [comment]
> 
> 
> IIRC _views_ was one of the ideas for what to call them originally (and was where the inspiration for "queries" came from). I don't remember why we decided not to go with it.
> 
> I'm saying you can use lua to filter to only tasks with `+a`, then use the left drawer to select `+b` and `+c`, in order to get tasks with `+a AND (+b OR +c)`. And then easily swap to `+a AND (+b OR +d)` if you want.

I see. So, I can't do a sorting-only filter without the whole package, or can lua do sorting too?

@mpcjanssen 
That's definitely disappointing, but I understand life can get in the way.

I really want to use Simpletask since it is clean and fast so I will hazard a question: does this "serious development" extend to something like alternating line backgrounds/two of preset colors for contexts and projects.

Thanks either way for the great app. 👍

**Comment by t214c** [comment]
> 
> 
> IIRC _views_ was one of the ideas for what to call them originally (and was where the inspiration for "queries" came from). I don't remember why we decided not to go with it.
> 
> I'm saying you can use lua to filter to only tasks with `+a`, then use the left drawer to select `+b` and `+c`, in order to get tasks with `+a AND (+b OR +c)`. And then easily swap to `+a AND (+b OR +d)` if you want.

@smichel17  I see. So, I can't do a sorting-only filter without the whole package, or can lua do sorting too?

@mpcjanssen 
That's definitely disappointing, but I understand life can get in the way.

I really want to use Simpletask properly since it is clean and fast so I will hazard a question: does this "serious development" extend to something like alternating line backgrounds (or just a separator line), two of preset colors for contexts and projects? Like [Markor](https://github.com/gsantner/markor) for example, but it doesn't have Simpletask's task manipulation powers.

![](https://raw.githubusercontent.com/gsantner/markor/master/metadata/en-US/phoneScreenshots/05.jpg)
Thanks either way for the great app. 👍

**Comment by smichel17** [comment]
@t214c Lua can do sorting, but it's not distinct from the filter. You can make a filter with some sorting, and then apply additional filtering on top of that, but you can't change the sorting while leaving the filtering the exact same.

The closest thing I can think of is: you can define some functions in the global lua config, and then call these from the filter. So then you can define several saved filters which all use the same sorting, but different filtering.
